### ISSUE TITLE
Attempt to read property “username” on null"

### ISSUE THREAD

**Issue by DikkiX** [promocat/yii2-twofa#2]
Hi, bij het installeren van deze extension voor mijn schoolproject loop ik tegen wat problemen aan.

Attempt to read property “username” on null

komt er tevoorschijn en nu ben ik even clueless hoe ik dit moet fixen. ik had ook een topic op de forum hierover gemaakt.
https://forum.yiiframework.com/t/2mfa-class-not-found/134701

**Comment by C4RoCKeT** [comment]
I'll reply in English, because this should be readable for everyone:

I'm sorry, but this is clearly a user error.
The error is telling you exactly what is wrong;
"user" is not set / null on your $model.

It's clear that you just started with and are still learning PHP, so I will
point you in the right direction.

You should not be rendering the QR code on the login page. In fact, a 2fa
code has nothing to do on the login page at all.

Let me put it this way;
How can you have a user with a username without the user actually
attempting to login.

The login process should be;
1. Regular login
2a. Check if user has 2fa enabled. If so; show login verification form (2fa
input)
2b; verify the 2fa code
3. Log the user in as normal.

The QR code is only used when enabling 2fa on an already existing user.

Try asking your class mates for help. And focus on the basics before
heavily depending on frameworks. Learning how to debug and understanding
the issue is 80% of the work ;)

Op di 20 dec. 2022 12:21 schreef DikkiX ***@***.***>:

**Comment by DikkiX** [comment]
Hi,

Thanks, i asked my class mates, but no one knows anything.
i am going to try and see how this works.
Is there by any chance an demo repository with this working so that i can check how this is implemented.
Unfortunately an schoolproject has made it an requirement to have this function without any lessons about it!

**Comment by C4RoCKeT** [comment]
We don't have a demo project. We use this package in one of our closed
sourced software projects.
(I'm replying using my private account to reply, but I do work for
PromoCat).

There is an example.php in the root. That combines with the basics I gave
you in my previous reply should give you a good start.

In a nutshell; leave the poor login view alone.

User has an account and wants to enable 2fa.
User gets shown a QR code that can be used with an authenticator app.
User enters a generated 2fa code for validation.
User has enabled 2fa.

See actionEnableTwoFa on line 67.

Create a view for enabling 2fa for an existing user. This is where you
render the QR widget.


Validate that the user has a totp_secret in the database.

Now for the logging in part:

Create a view for entering the 2fa code (login verification) and;

Line 20: check if the user has 2fa enabled

Line 24: create the login verification session.
In other words: remember the user that is trying to login, but prompt them
for a 2fa code in the next step.

Line 25: redirect to actionLoginVerification on line 34.

Line 39: retrieve the user from the session we created on line 24.

Line 49: load the post data and attempt to login the user (verify the given
2fa code).

---

This example was last updated a little over 4,5 years ago and could be a
little bit out of date. It's an example and should not be considered
anything more.




Op di 20 dec. 2022 13:12 schreef DikkiX ***@***.***>:
